plugins/statistics.koplugin/main.lua

```diff
@@ -598,6 +603,38 @@ function ReaderStatistics:upgradeDBto20201022(conn)
     conn:exec("PRAGMA user_version=20201022;")
 end
 
+function ReaderStatistics:upgradeDBto20221022(conn)
+    conn:exec([[
+        -- We make the index on book's (title, author, md5) unique in order to sync dbs
+        -- First we remove duplicates (most likely none existant?)
+        UPDATE page_stat_data SET id_book = (
+            SELECT map.min_id FROM (
+                SELECT id, (
+                    SELECT min(id) FROM book b2
+                    WHERE (book.title, book.authors, book.md5) = (b2.title, b2.authors, b2.md5)
+                ) as min_id
+                FROM book WHERE book.id >= min_id
+            ) as map WHERE page_stat_data.id_book = map.id
+        );
+        DELETE FROM book WHERE id > (
+            SELECT MIN(id) FROM book b2
+            WHERE (book.title, book.authors, book.md5) = (b2.title, b2.authors, b2.md5)
+        );
```

Comment:
I started a review, HORRIFIED, thinking these would `update page_stat_date set id_book=<a single fixed constant id_book>` and `delete all book where id > <a single book id>`, seeing all the rows where touched.
But no, it works, and I learned something about SQL :) that the update and the delete will be done row by row checking the condition that follows for the values of each row.
(I had also never seen these tuple comparisons `(a,b,c) = (a,b,c)`, I was used to using `a=a, b=b, c=c` :)

Anyway, you might have an issue when one of authors, title, md5 is NULL.
```
sqlite> select title, authors, md5, count(1) nb from book group by title, authors, md5 having nb > 1;
title      authors  md5                               nb
---------  -------  --------------------------------  --
djvu3spec  NulL     99c211540c6b2d46441d8f45347ccbb5  13
sqlite> select * from book where md5='99c211540c6b2d46441d8f45347ccbb5';
id    title      authors  notes  last_open   highlights  pages  series  language  md5                               total_read_time  total_read_pages
----  ---------  -------  -----  ----------  ----------  -----  ------  --------  --------------------------------  ---------------  ----------------
1198  djvu3spec  NulL     0      1585337100  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  12               1
1199  djvu3spec  NulL     0      1585337182  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  41               1
1200  djvu3spec  NulL     0      1585337195  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  10               1
1201  djvu3spec  NulL     0      1585337203  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  5                1
1202  djvu3spec  NulL     0      1585337215  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  NulL             0
1203  djvu3spec  NulL     0      1585337236  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  14               1
1204  djvu3spec  NulL     0      1585337263  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  22               1
1205  djvu3spec  NulL     0      1585337271  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  5                1
1206  djvu3spec  NulL     0      1585337280  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  5                1
1207  djvu3spec  NulL     0      1585338561  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  NulL             0
1250  djvu3spec  NulL     0      1588543107  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  27               1
1252  djvu3spec  NulL     0      1588666043  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  45               1
1253  djvu3spec  NulL     0      1588666589  0           71     NulL    NulL      99c211540c6b2d46441d8f45347ccbb5  6                1
```
The page_stat_data from these books seem to have disappear.
(Not sure why I have these duplicates. Might be something with djvu I never have nor read. I also remember we had a but with statistics with md5 leaking and possibly these multiple entries. We solved the bug, I don't know if we also fixed the DB (and may be we didn't fix the above because of the NULL authors.)
(These strange `NulL` is me having in my .sqliterc: `.nullvalue NulL` - there are really NULL values.)

Reply:
After some digging, I think the most practical way to solve the `null` problem is make the columns `NOT NULL DEFAULT ""` (we are modifying the columns anyway)

Reply:
Or I can keep the `NULL` values by adding a bunch `ifnull()` functions and change some `=` to `IS`. It is a bit messier but doesn't add, update, drop, rename columns which I'm somewhat scared of.

Reply:
Well, after some more digging it seems we can just leave things be: after #7868 there wouldn't be any more nulls. The ones before are ignored anyway.

Reply:
> Well, after some more digging it seems we can just leave things be: after https://github.com/koreader/koreader/pull/7868 there wouldn't be any more nulls. The ones before are ignored anyway.

What do you mean ?
Dunno if users may get books more serious than my djvu3spec with their authors (or titles?) field NULL, but it seems the associated page_stat_data just got removed - which should make them no longer visible in Calendar view.

May be you should just as part of the migration do first: `update authors='' where authors is NULL` ?

Reply:
Will add said update in migration. 
I'm not 100% sure though, but looking at the code for getting book id, it seems null author book will never be selected, since `Null = Null` is false, nor `Null = 'N/A'`? Or maybe the `prepare` then `bind` can make it happen...
```
sql_stmt = [[
SELECT id
FROM   book
WHERE  title = ?
    AND authors = ?
    AND md5 = ?;
]]
stmt = conn:prepare(sql_stmt)
result = stmt:reset():bind(self.data.title, self.data.authors, self.data.md5):step()
id_book = result[1]
```

Reply:
> the code for getting book id, it seems null author book will never be selected

By this kind of SQL query, may be.
But there are possibly other kind of queries that don't "get" by authors (ie. Stats per months, Calendar, where we just "list" things, so from page_stat_data in the period > book_id > title/authors for display)

Reply:
Speaking of Calendar, I have a small request, which is not that big a deal really, but the color scheme is a little unbalanced. Once in a while, a bar gets pure black and really stands out among others. It attracts unnecessary(?) attention and implies something special about a book(could it be that there's really something special?). Maybe lessen a bit on the darker end of the spectrum?

Reply:
I agree the white on black one is a bit too striking.
Dunno if you could think of one other combination that would be different enough from the 7 others to replace it - or just get rid of it. We have to use these named values, as we're limited to the 8 colors available on eInk (any other will show as one of these 8).
https://github.com/koreader/koreader/blob/4d8e2f0ea15d15f8ce973fc7d5634757bea3eb07/plugins/statistics.koplugin/calendarview.lua#L247-L257

Reply:
There are 16 shades of gray, actually.

The named constants are rather abritrary, both in their amount, and in which value they use ;).

(It was mostly a case of "hey, that one looks ggood here, and I'm using it more that once").

`BlitBuffer.Color8(0x00)` (BLACK) to `BlitBuffer.Color8(0xFF)` white, and the stride is `0x11`, which makes the palette very easy to remember ;).

Reply:
>(any other will show as one of these 8)

Theoretically we can have 16 degrees of gray as in 0x00, 0x11 ... 0xff? Maybe name a couple more? https://github.com/koreader/koreader-base/pull/853#discussion_r265480883

Reply:
For reference
![1](https://user-images.githubusercontent.com/62179190/198889135-0f3ea9e9-4949-47f2-a782-a95d96e8729e.png)

Reply:
Some grays are not that different :) Even if more inserted, the effect may not be too different than say just use four colors. Don't know the actual logic for deciding bar color though.

Reply:
Go ahead with 16 if needed.
No specific logic, I just thought we could have 8 different calendar span colors because why not, 4 might be too few (the less difference there are, the more different books will get the same color, which might be confusing when you follow them from end of a week to begin of the next week, if you end a book a sunday, open 3 books that sunday, and start another on monday, and they get the same color).

Reply:
Then to avoid the striking black, moving the Dim_Gray down to last and adding another one in its old place seem okay.

Reply:
I’d like to recommend b/w textures. 
The application would provide more distinctiveness. 
![4567CEFE-5A2C-4A75-AB40-7743F774ED23](https://user-images.githubusercontent.com/95706893/198915350-809217dd-43ac-43b9-b675-b94dbfc64dcc.png)

Reply:
Checkerboard patterns are... very bad for, and on, eInk ;).

Reply:
It doesn’t have to be that texture. Stripes could be sufficient. To fulfill the need.

Reply:
Keep in mind we're talking about very thin bars *with* text on top of it: it would be awfully unreadable ;).

Reply:
Ok. Thought it might help the no color problem. Guess I was wrong.